embassy-stm32/src/spi/spi_v1.rs

```diff
@@ -0,0 +1,364 @@
+                // spin
+            }
+            unsafe {
+                regs.dr().write(|reg| reg.0 = *word as u32);
+            }
+            while unsafe { !regs.sr().read().rxne() } {
+                // spin waiting for inbound to shift in.
+            }
+            *word = unsafe { regs.dr().read().0 as u8 };
+            let sr = unsafe { regs.sr().read() };
+            if sr.fre() {
+                return Err(Error::Framing);
+            }
+            if sr.ovr() {
+                return Err(Error::Overrun);
+            }
+            if sr.crcerr() {
+                return Err(Error::Crc);
+            }
+        }
+
+        Ok(words)
+    }
+}
+
+impl<'d, T: Instance> embedded_hal::blocking::spi::Write<u16> for Spi<'d, T> {
+    type Error = Error;
+
+    fn write(&mut self, words: &[u16]) -> Result<(), Self::Error> {
+        self.set_word_size(WordSize::SixteenBit);
+        let regs = T::regs();
+
+        for word in words.iter() {
+            while unsafe { !regs.sr().read().txe() } {
+                // spin
+            }
+            unsafe {
+                regs.dr().write(|reg| reg.0 = *word as u32);
+            }
+            loop {
+                let sr = unsafe { regs.sr().read() };
+                if sr.fre() {
+                    return Err(Error::Framing);
+                }
+                if sr.ovr() {
+                    return Err(Error::Overrun);
+                }
+                if sr.crcerr() {
+                    return Err(Error::Crc);
+                }
+                if !sr.txe() {
+                    // loop waiting for TXE
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<'d, T: Instance> embedded_hal::blocking::spi::Transfer<u16> for Spi<'d, T> {
+    type Error = Error;
+
+    fn transfer<'w>(&mut self, words: &'w mut [u16]) -> Result<&'w [u16], Self::Error> {
+        self.set_word_size(WordSize::SixteenBit);
+        let regs = T::regs();
+
+        for word in words.iter_mut() {
+            while unsafe { !regs.sr().read().txe() } {
+                // spin
+            }
+            unsafe {
+                regs.dr().write(|reg| reg.0 = *word as u32);
+            }
+            while unsafe { !regs.sr().read().rxne() } {
+                // spin waiting for inbound to shift in.
+            }
+            *word = unsafe { regs.dr().read().0 as u16 };
+            let sr = unsafe { regs.sr().read() };
+            if sr.fre() {
+                return Err(Error::Framing);
+            }
+            if sr.ovr() {
+                return Err(Error::Overrun);
+            }
+            if sr.crcerr() {
+                return Err(Error::Crc);
+            }
+        }
+
+        Ok(words)
+    }
+}
+
+
+pub(crate) mod sealed {
+    use super::*;
+    use embassy::util::AtomicWaker;
+
+    pub trait Instance {
+        fn regs() -> &'static spi::Spi;
+    }
+
+    pub trait Sck<T: Instance>: Pin {
```

Comment:
(I know this was added in an earlier PR, but) the naming for these traits is inconsistent with USART's. MosiPin vs Mosi. Could you name then FooPin?